### PR TITLE
More ocr models

### PR DESCRIPTION
**Pull request by AllentDan**
## Motivation

Add more mmocr models. PANet, PSENet, SATRN.

## Checklist

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x]  The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: cmake/modules/FindTENSORRT.cmake [hunk 1]
```diff
@@ -9,22 +9,24 @@ endif()
...
     TENSORRT_INCLUDE_DIR NvInfer.h
-    HINTS ${TENSORRT_DIR}
+    HINTS ${TENSORRT_DIR} ${CUDA_TOOLKIT_ROOT_DIR}
     PATH_SUFFIXES include)
 
 if (NOT TENSORRT_INCLUDE_DIR)
-    message(FATAL_ERROR "Cannot find TensorRT header NvInfer.h, "
-        "please check if the path is correct")
+    message(FATAL_ERROR "Cannot find TensorRT header NvInfer.h "
+        "in TENSORRT_DIR: ${TENSORRT_DIR} or in CUDA_TOOLKIT_ROOT_DIR: "
```
Comment (grimoire) [inline cmake/modules/FindTENSORRT.cmake:17]: Why do we search TensorRT in `CUDA_TOOLKIT_ROOT_DIR`?

File: mmdeploy/backend/tensorrt/utils.py [hunk 2]
```diff
@@ -80,6 +81,8 @@ def create_trt_engine(onnx_model: Union[str, onnx.ModelProto],
         builder.max_workspace_size = max_workspace_size
 
     config = builder.create_builder_config()
+    if set_tactic_sources:
```
Comment (grimoire) [inline mmdeploy/backend/tensorrt/utils.py:84]: `set_tactic_sources` looks like a function name and the behavior is strange.
`TacticSource.CUBLAS` is only used on cuda10.2 without patch. Is it worth to add a flag for it?

File: mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (grimoire) [inline mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py:1]: Does not need to add `trt_fp16` in the file name even when your are solving that problem.

**Comment by grimoire** [comment]
`CRNN` failed because the input shape is `[1,1,32,32]`.
visualize failed on latest MMOCR. The visualize function is static method now.

File: mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    Args:
+        ctx (ContextCaller): The context with additional information.
+        self: The instance of the class FPEM_FFM.
+        x (Tensor): The input tensor of shape (N, C, H, W).
+
+    Returns:
+        outs (Tensor): The output tensor of shape (N, C, H, W).
+    """
+    if self.conv1.in_channels < CHANNEL_THRESH:
```
Comment (grimoire) [inline mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py:107]: It the weight is large, it will still overflow whatever CHANNEL_THRESH is.

File: mmdeploy/pytorch/functions/triu.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+
+from mmdeploy.core import FUNCTION_REWRITER
+
+
+@FUNCTION_REWRITER.register_rewriter(func_name='torch.triu')
+def triu(ctx,
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/triu.py:8]: `triu` support batched input.

File: mmdeploy/pytorch/functions/triu.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+         **kwargs) -> torch.Tensor:
+    """Rewrite `triu` for exporting model to ONNX."""
+    size = input.shape[0]
+    arange = torch.arange(size, device=input.device)
+    mask = arange.expand(size, size)
+    arange = arange.unsqueeze(-1)
+    if diagonal:
+        arange = arange + diagonal
+    mask = mask >= arange
+    return input.masked_fill(mask == 0, 0)
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/triu.py:21]: Since mask is a bool tensor, why not use `~mask`.

File: mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+    identity = x
+
+    out = self.conv1(x)
+    out = self.norm1(out)
+    out = self.relu(out)
+
+    out = self.conv2(out)
+
+    if torch.abs(self.norm2(out)).max() < 65504:
```
Comment (grimoire) [inline mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py:118]: Note that this is the result of the example input. The result model might still overflow on real data.

File: cmake/modules/FindTENSORRT.cmake (hunk 1 above)
Comment (AllentDan) [reply cmake/modules/FindTENSORRT.cmake:17]: It was brought from master to dev-0.5.0

File: mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py (hunk 4 above)
Comment (AllentDan) [reply mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py:107]: CHANNEL_THRESH is only used for filtering out the last basic_block of resnet.

**Comment by AllentDan** [comment]
> 

CRNN requires other configs with 1x32x32 specified. I'm actually not sure if we should keep the codes for fp-16 because it only works for one checkpoint.

File: docs/en/benchmark.md [hunk 8]
```diff
@@ -1338,6 +1338,80 @@ Users can directly test the performance through [how_to_evaluate_a_model.md](tut
...
+    <td align="center">recall</td>
+    <td align="center">0.7401</td>
+    <td align="center">0.7401</td>
+    <td align="center">0.7401</td>
+    <td align="center">0.7357</td>
+    <td align="center">0.7366</td>
+    <td align="center">-</td>
+    <td align="center">-</td>
+    <td align="center">0.7401</td>
+    <td align="center" rowspan="3">$MMOCR_DIR/configs/textdet/panet/panet_r18_fpem_ffm_600e_icdar2015.py</td>
```
Comment (RunningLeon) [inline docs/en/benchmark.md:1391]: may remove `align="center"` attribute for config column.

File: docs/zh_cn/03-benchmark/benchmark.md [hunk 9]
```diff
@@ -992,6 +992,78 @@ GPU: ncnn, TensorRT, PPLNN
...
+    <td align="center">0.8057</td>
+    <td align="center">0.8057</td>
+    <td align="center">0.8057</td>
+    <td align="center">0.8054</td>
+    <td align="center">0.7989</td>
+    <td align="center">-</td>
+    <td align="center">0.8057</td>
+  </tr>
+  <tr>
+    <td align="center" rowspan="3"><a href="https://github.com/open-mmlab/mmocr/tree/master/configs/textdet/panet/panet_r18_fpem_ffm_600e_icdar2015.py">PANet</a></td>
```
Comment (RunningLeon) [inline docs/zh_cn/03-benchmark/benchmark.md:1032]: may remove `align="center"` for config column and use default left alignment setting.

**Comment by lvhan028** [comment]
Hi, @AllentDan may use dev-v0.5.0 as your base branch on this PR.

File: mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py (hunk 3 above)
Comment (RunningLeon) [reply mmdeploy/codebase/mmocr/models/text_detection/panet_trt_fp16.py:1]: maybe we should move this part to document as one of reference solution since this rewriting may only work on a specific checkpoint.

**Comment by lzhangzz** [comment]
The exported pipeline.json for PANet has both `size` and `img_scale` set
```JSON
{
    "type": "Resize",
    "img_scale": [
        3000,
        640
    ],
    "keep_ratio": true,
    "size": [
        736,
        1333
    ]
}
```

**Comment by lzhangzz** [comment]
The exported pipeline.json for PANet has both `size` and `img_scale` set
```JSON
{
    "type": "Resize",
    "img_scale": [
        3000,
        640
    ],
    "keep_ratio": true,
    "size": [
        736,
        1333
    ]
}
```

File: mmdeploy/pytorch/functions/triu.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    # statically conversion
+    if height <= width:
+        arange = arange[:height]
+    else:
+        arange = torch.cat(
+            [arange, arange.new_zeros(height - width, 1) + width])
+    if diagonal:
+        arange = arange + diagonal
+    mask = mask >= arange
+    mask.expand(input.shape)
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/triu.py:28]: Expand is not a inplace operation. This statement did nothing. Can be removed.

File: mmdeploy/pytorch/functions/triu.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+def triu(ctx,
+         input: torch.Tensor,
+         diagonal: int = 0,
+         *args,
+         **kwargs) -> torch.Tensor:
+    """Rewrite `triu` for exporting model to ONNX."""
+    assert len(input.shape) >= 2
+    height, width = input.shape[-2:]
+    arange = torch.arange(width, device=input.device)
+    mask = arange.expand(height, width)
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/triu.py:17]: You can use `mask = arange.unsqueeze(0)`.
The binary broadcast can do everything else for you.

File: mmdeploy/pytorch/functions/triu.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    arange = arange.unsqueeze(-1)
+    # statically conversion
+    if height <= width:
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/triu.py:20]: Not safe for dynamic input shape.
the relation between height and width might change.

Comment (AllentDan) [reply mmdeploy/pytorch/functions/triu.py:20]: Yes, I didn't come up with any solution for this yet.

File: mmdeploy/pytorch/functions/triu.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+@FUNCTION_REWRITER.register_rewriter(func_name='torch.triu')
+def triu(ctx,
+         input: torch.Tensor,
+         diagonal: int = 0,
+         *args,
+         **kwargs) -> torch.Tensor:
+    """Rewrite `triu` for exporting model to ONNX."""
+    assert len(input.shape) >= 2
+    height, width = input.shape[-2:]
+    arange = torch.arange(width, device=input.device)
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/triu.py:16]: I think:
```python
    arange0 = torch.arange(width, device=input.device).unsqueeze(0)
    arange1 = torch.arange(height, device=input.device).unsqueeze(-1)
    mask = arange0 >= arange1
    return input * mask
```
is all you need.

**Review by grimoire — APPROVED** [review]
LGTM

File: docs/zh_cn/03-benchmark/benchmark.md (hunk 9 above)
Comment (AllentDan) [reply docs/zh_cn/03-benchmark/benchmark.md:1032]: `align="center"` is aligned with other settings of Chinese benchmark.

**Review by RunningLeon — DISMISSED** [review]
LGTM

**Review by lzhangzz — APPROVED** [review]
LGTM

**Review by grimoire — APPROVED** [review]
